lib/wit/dependency.py

```diff
@@ -0,0 +1,155 @@
+                                              source_map[subdep.name]))
+                    sys.exit(1)
+
+            source_map[subdep.name] = subdep.package.source
+
+            if subdep.package.repo is None:
+                continue
+
+            if subdep.get_commit_time() > self.get_commit_time():
+                log.error("Repo [{}] has a dependent that is newer than the source. "
+                          "This should not happen.\n".format(subdep.name))
+                sys.exit(1)
+
+            commit_time = subdep.get_commit_time()
+            queue.append((commit_time, subdep))
+
+        queue.sort(key=lambda tup: tup[0])
+
+        return source_map, packages, queue
+
+    def __key(self):
+        return (self.source, self.specified_revision, self.name)
+
+    def __hash__(self):
+        return hash(self.__key())
+
+    def __eq__(self, other):
+        return isinstance(self, type(other)) and self.__key() == other.__key()
+
+    @staticmethod
+    def infer_name(source):
+        return Path(source).name.replace('.git', '')
+
+    # wsroot, repo_paths, packages, force_root
+    def load_package(self, packages, repo_paths):
+        """
+        Bind itself to a package, using one in the `packages` argument if available.
+        This will also add itself as a dependent of the package
+        """
+        if self.name in packages:
+            self.package = packages[self.name]
+        else:
+            self.package = Package(self.name, self.source, self.specified_revision, repo_paths)
+        self.package.add_dependent(self)
+
+    def add_dependent(self, dependent):
+        if dependent not in self.dependents:
+            self.dependents.append(dependent)
+
+    def get_commit_time(self):
+        return datetime.utcfromtimestamp(int(self.package.repo.commit_to_time(
+            self.specified_revision)))
+
+    def manifest(self):
+        return {
+            'name': self.name,
+            'source': self.source,
+            'commit': self.specified_revision,
+        }
+
+    # used before saving to manifests/lockfiles
+    def resolved(self):
+        if self.package is None or self.package.repo is None:
+            raise Exception("Cannot resolve dependency that us unbound to disk")
+        return Dependency(self.name, self.source, self.resolved_rev())
+
+    def resolved_rev(self):
+        if self.package.repo is None or self.package.repo is None:
+            raise Exception("Cannot resolve dependency that is unbound to disk")
+        return self.package.repo.get_commit(self.specified_revision)
+
+    def __repr__(self):
+        return "Dep({})".format(self.tag())
+
+    def short_revision(self):
+        if self.package and self.package.repo:
+            if self.package.repo.is_hash(self.specified_revision):
+                return self.package.repo.get_shortened_rev(self.specified_revision)
+            return self.specified_revision
+        return self.specified_revision[:8]
```

Comment:
The reason I added the `[:8]` here is because, when doing `wit inspect --tree`, the dependency tag is requested before the repo is loaded. Therefore, without the repo loaded, the Dependency cannot get the shortest unique string representing its hash.

This only applies to hashes before arrows in the tree output. Lone hashes (e.g. `foo:2891731`) and replacement hashes (`->19371239`) are still based on `git rev-parse --short`.

Should we just cut it off at 8 or should we try loading repos earlier in the process of `wit inspect --tree`?